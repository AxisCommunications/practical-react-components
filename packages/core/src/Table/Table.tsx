import React, {
  useRef,
  createContext,
  useContext,
  useMemo,
  useReducer,
  Reducer,
  useLayoutEffect,
  useState,
} from 'react'
import styled from 'styled-components'

import { spacing } from '../designparams'

import { ColumnResizerRow } from './ColumnResizerRow'
import { useResetScroll } from 'react-hooks-shareable'
import { SCROLLBAR_DIMENSION } from '../Global/GlobalScrollbarStyle'

type BaseElement = HTMLDivElement
type BaseProps = React.HTMLAttributes<BaseElement>

const PADDING_LEFT = 24

export const TABLE_DIMENSIONS = {
  DEFAULT_MIN_COLUMN_WIDTH: 64,
  PADDING_LEFT,
  PADDING: PADDING_LEFT + SCROLLBAR_DIMENSION, // total padding (sum of left and right(scrollbar))
  SELECT_WIDTH: 48, // holds the checkbox
  MENU_WIDTH: 32 + (20 - SCROLLBAR_DIMENSION), // holds the menu (24px box that holds menu button icon should have 24px space between itself and Table edge)
  // TODO: this is componentSize, but that one is a string and contains 'px'
  ROW_HEIGHT: 56,
}

/*******************************************************************************
 *
 * Column width control.
 *
 * The columns of a table always have a fixed width,
 * which is set initially either by the user or by
 * detecting the rendered size of the table, and dividing
 * it by the available columns. When you specify your
 * own initial widths, they will be scaled to the table
 * size automatically (and when resizing).
 *
 * The array of widths is updated using a reducer.
 *
 ******************************************************************************/

export enum WidthActionType {
  SET_WIDTHS,
  UPDATE_WIDTHS,
  RESET_WIDTHS,
  SET_TOTAL_WIDTH,
}

interface ISetWidthsAction {
  readonly type: WidthActionType.SET_WIDTHS
  readonly widths: ReadonlyArray<number>
}

interface IUpdateWidthsAction {
  readonly type: WidthActionType.UPDATE_WIDTHS
  readonly widths: { readonly [id: number]: number | undefined }
  readonly onWidthsChange?: (widths: ReadonlyArray<number>) => void
}

interface IResetWidthsAction {
  readonly type: WidthActionType.RESET_WIDTHS
  readonly numberOfColumns: number
}

interface ISetTotalWidthAction {
  readonly type: WidthActionType.SET_TOTAL_WIDTH
  readonly value: number
}

type WidthAction =
  | ISetWidthsAction
  | IUpdateWidthsAction
  | IResetWidthsAction
  | ISetTotalWidthAction

interface IWidthsState {
  readonly totalWidth: number
  readonly columnWidths: ReadonlyArray<number>
  readonly initialWidths: ReadonlyArray<number>
  readonly minColumnWidth: number
}

const reduceWidths = (
  state: IWidthsState,
  action: WidthAction
): IWidthsState => {
  switch (action.type) {
    case WidthActionType.SET_WIDTHS: {
      // Sanitize columns to the minimum width
      const sanitizedWidths = action.widths.map(width =>
        Math.max(state.minColumnWidth, width)
      )
      const totalColumnWidths = sanitizedWidths.reduce(
        (acc, cur) => acc + cur,
        0
      )
      const normalizedWidths = sanitizedWidths.map(
        width => width / totalColumnWidths
      )
      // Scale all columns to fill the total width.
      return {
        ...state,
        columnWidths: normalizedWidths.map(normalizedWidth =>
          Math.max(state.minColumnWidth, normalizedWidth * state.totalWidth)
        ),
      }
    }
    case WidthActionType.UPDATE_WIDTHS: {
      const columnWidths = state.columnWidths.map((value, id) => {
        const width = action.widths[id]
        return width !== undefined
          ? Math.max(state.minColumnWidth, width)
          : value
      })
      if (action.onWidthsChange !== undefined) {
        action.onWidthsChange(columnWidths)
      }
      return {
        ...state,
        columnWidths,
      }
    }
    case WidthActionType.RESET_WIDTHS: {
      const columnWidth =
        action.numberOfColumns !== 0
          ? state.totalWidth / action.numberOfColumns
          : 0
      return {
        ...state,
        columnWidths: new Array(action.numberOfColumns).fill(
          Math.max(state.minColumnWidth, columnWidth)
        ),
      }
    }
    case WidthActionType.SET_TOTAL_WIDTH: {
      const totalWidth = action.value

      // We always base the calculation on the initial widths if possible,
      // in order to keep the widths stable upon resize.
      const columnWidths =
        state.initialWidths.length !== 0
          ? state.initialWidths
          : state.columnWidths

      const totalColumnWidths = columnWidths.reduce((acc, cur) => acc + cur, 0)
      const nColumns = columnWidths.length

      if (totalWidth === totalColumnWidths || nColumns === 0) {
        return {
          ...state,
          totalWidth,
        }
      }

      const minWidth = nColumns * state.minColumnWidth

      const availableWidth = totalWidth - minWidth
      if (availableWidth <= 0) {
        // There is no available width, all columns get the minimum width
        return {
          ...state,
          totalWidth,
          columnWidths: columnWidths.map(() => state.minColumnWidth),
        }
      }

      // Distribute available width to all columns
      const sanitizedWidths = columnWidths.map(width =>
        Math.max(state.minColumnWidth, width)
      )
      const sanitizedTotalWidth = sanitizedWidths.reduce(
        (acc, cur) => acc + cur,
        0
      )
      const extraColumnWidth = sanitizedTotalWidth - minWidth

      if (extraColumnWidth <= 0) {
        // There is no extra width, all columns get the same new width
        return {
          ...state,
          totalWidth,
          columnWidths: columnWidths.map(() => totalWidth / nColumns),
        }
      }

      const normalizedWidths = sanitizedWidths.map(
        width => Math.max(width - state.minColumnWidth) / extraColumnWidth
      )

      return {
        ...state,
        totalWidth,
        columnWidths: normalizedWidths.map(
          normalizedWidth =>
            state.minColumnWidth + normalizedWidth * availableWidth
        ),
      }
    }
    default: {
      return state
    }
  }
}

/*******************************************************************************
 *
 * TableContext
 *
 * Keeps track of table-wide properties which need to
 * be accessible from within each of the subcomponents.
 *
 ******************************************************************************/

export const TableContext = createContext<{
  readonly minColumnWidth: number
  readonly columnWidths: ReadonlyArray<number>
  readonly dispatchWidthsAction: React.Dispatch<WidthAction>
  readonly selectWidth: number
  readonly menuWidth: number
  readonly onSelect?: (selected: boolean, id?: string) => void
  readonly hasMenu: boolean
  readonly onWidthsChange?: (widths: ReadonlyArray<number>) => void
}>({
  minColumnWidth: 0,
  columnWidths: [],
  dispatchWidthsAction: () => {
    /** */
  },
  selectWidth: 0,
  menuWidth: 0,
  hasMenu: false,
  onWidthsChange: () => {
    /** */
  },
})

/**
 * Transform the array of widths to an appropriate grid-template-columns
 * string, taking into account select/menu columns.
 *
 * Example:
 *  [34, 78]
 */
export const useGridTemplateColumns = () => {
  const { columnWidths, selectWidth, menuWidth } = useContext(TableContext)

  // Set up grid template columns from widths
  const gridTemplateColumnsStyle = useMemo(() => {
    return {
      gridTemplateColumns: [
        `${TABLE_DIMENSIONS.PADDING_LEFT}px`,
        '[start]',
        `${selectWidth}px`,
        '[content-start]',
        ...columnWidths.map(columnWidth => `${columnWidth}px`),
        '[content-end]',
        `${menuWidth}px`,
        '[scroll]',
        `${SCROLLBAR_DIMENSION}px`,
        '[end]',
      ].join(' '),
    }
  }, [columnWidths, menuWidth, selectWidth])

  return gridTemplateColumnsStyle
}

/*******************************************************************************
 *
 * Internal Table building blocks.
 *
 ******************************************************************************/

/**
 * The basic components of our grid.
 *
 * Note: grid-template-columns is set using a style attribute
 * to allow for dynamic change (without affecting the class name)
 */

const TableContainer = styled.div<{
  readonly dragging: boolean
}>`
  overflow-x: auto;
  overflow-y: hidden;
  position: relative;
  pointer-events: ${({ dragging }) => (dragging ? 'none' : undefined)};
  height: 100%;
  width: 100%;
  background-color: ${({ theme }) => theme.color.background00()};
`

/**
 * TableHeaderGrid
 *
 * A grid containing the header cells.
 */
const TableHeaderContainer = styled.div`
  display: grid;
  width: 100%;
  border-bottom: 1px solid ${({ theme }) => theme.color.element12()};
`

/**
 * TableContentGrid
 *
 * A grid containing the content cells.
 */
const TableContentContainer = styled.div.attrs<{
  readonly width: number
  readonly maxHeight: number
}>(({ width, maxHeight }) => {
  return {
    style: { width: `${width}px`, maxHeight: `${maxHeight}px` },
  }
})<{
  readonly maxHeight?: number
  readonly width: number
}>`
  display: grid;
  overflow-y: auto;
  overflow-x: hidden;
  padding-bottom: ${spacing.medium};
  position: relative;
  z-index: 0;
`

/**
 * Table
 *
 * Defined by the number of columns.
 */

interface ITableProps extends Omit<BaseProps, 'onSelect'> {
  /**
   * `class` to be passed to the component.
   */
  readonly className?: BaseProps['className']
  /**
   * The initial widths for the columns of the Table.
   * The width will be automatically adjusted to fit the
   * table, so use 0 if not known, e.g. new Array(#columns).fill(0)
   * would be a valid starting point.
   */
  readonly initialWidths?: ReadonlyArray<number>
  /**
   * Control if columns should be resizeable or not. If true, resize
   * handles will be available to change column width.
   *
   * @default false
   */
  readonly resizableColumns?: boolean
  /**
   * Minimum column width.
   *
   * @default DEFAULT_MIN_COLUMN_WIDTH
   */
  readonly minColumnWidth?: number
  /**
   * Callback which will get the (new) array of column widths
   * every time any width is changed.
   */
  readonly onWidthsChange?: (widths: ReadonlyArray<number>) => void
  /**
   * The maximum height of the table in number of rows,
   * more rows will cause a scrollbar to appear.
   */
  readonly maxHeight?: number
  /**
   * Callback which will get checked status and row ID on selection.
   * The ID is undefined if the main checkbox is selected ("select all").
   */
  readonly onSelect?: (selected: boolean, id?: string) => void
  /**
   * Allow for optional right-side menu content for each row.
   */
  readonly hasMenu?: boolean
  /**
   * Key to control scrollbar in Table.
   * If key changes, scrollbar will scroll to top.
   */
  readonly scrollKey?: string
}

const EMPTY_ARRAY: ReadonlyArray<number> = []

export const Table: React.FunctionComponent<ITableProps> = React.memo(
  ({
    initialWidths = EMPTY_ARRAY,
    resizableColumns = false,
    minColumnWidth = TABLE_DIMENSIONS.DEFAULT_MIN_COLUMN_WIDTH,
    onWidthsChange,
    maxHeight,
    onSelect,
    hasMenu = false,
    scrollKey,
    children,
    ...props
  }) => {
    // Table width/height, including column widths
    const [{ columnWidths }, dispatchWidthsAction] = useReducer<
      Reducer<IWidthsState, WidthAction>
    >(reduceWidths, {
      totalWidth: initialWidths.reduce((sum, width) => sum + width, 0),
      columnWidths: initialWidths,
      initialWidths,
      minColumnWidth,
    })
    const [tableHeight, setTableHeight] = useState<number>()

    // Keep track of additional columns for select/menu.
    const selectWidth = useMemo(
      () => (onSelect !== undefined ? TABLE_DIMENSIONS.SELECT_WIDTH : 0),
      [onSelect]
    )
    const menuWidth = useMemo(
      () => (hasMenu ? TABLE_DIMENSIONS.MENU_WIDTH : 0),
      [hasMenu]
    )

    // Set up total width control
    const tableRef = useRef<HTMLDivElement>(null)

    useLayoutEffect(() => {
      const tableEl = tableRef.current
      if (tableEl !== null) {
        const updateTableDimensions = () => {
          const { width, height } = tableEl.getBoundingClientRect()
          dispatchWidthsAction({
            type: WidthActionType.SET_TOTAL_WIDTH,
            value: width - selectWidth - menuWidth - TABLE_DIMENSIONS.PADDING,
          })
          setTableHeight(height)
        }
        updateTableDimensions()

        const observer = new window.ResizeObserver(updateTableDimensions)
        observer.observe(tableEl)

        return () => observer.disconnect()
      }
    }, [columnWidths.length, selectWidth, menuWidth, setTableHeight])

    const [dragging, setDragging] = useState(false)

    /**
     * The content's width and (maximum) height.
     */
    const contentWidth = useMemo(() => {
      return (
        columnWidths.reduce((total, width) => total + width, 0) +
        selectWidth +
        menuWidth +
        TABLE_DIMENSIONS.PADDING
      )
    }, [columnWidths, selectWidth, menuWidth])
    const contentHeight =
      maxHeight !== undefined
        ? maxHeight * TABLE_DIMENSIONS.ROW_HEIGHT
        : tableHeight !== undefined
        ? tableHeight - TABLE_DIMENSIONS.ROW_HEIGHT
        : undefined

    const { header, rows } = useMemo(() => {
      const [headerEl, ...rowsEl] = React.Children.toArray(children)
      return { header: headerEl, rows: rowsEl }
    }, [children])

    const tableContentRef = useRef<HTMLDivElement>(null)
    // Scroll to top when scrollKey changes
    useResetScroll(tableContentRef, scrollKey)

    return (
      <TableContainer ref={tableRef} dragging={dragging} {...props}>
        <TableContext.Provider
          value={{
            minColumnWidth,
            columnWidths,
            dispatchWidthsAction,
            selectWidth,
            menuWidth,
            onSelect,
            hasMenu,
            onWidthsChange,
          }}
        >
          <TableHeaderContainer>{header}</TableHeaderContainer>
          <TableContentContainer
            maxHeight={contentHeight}
            width={contentWidth}
            ref={tableContentRef}
          >
            {tableHeight !== undefined ? rows : undefined}
          </TableContentContainer>
          {resizableColumns ? (
            <ColumnResizerRow setDragging={setDragging} />
          ) : null}
        </TableContext.Provider>
      </TableContainer>
    )
  }
)
