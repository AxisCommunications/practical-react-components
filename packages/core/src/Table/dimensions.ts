import { SCROLLBAR_DIMENSION } from '../Global/GlobalScrollbarStyle'

const PADDING_LEFT = 24

export const TABLE_DIMENSIONS = {
  DEFAULT_MIN_COLUMN_WIDTH: 64,
  PADDING_LEFT,
  PADDING: PADDING_LEFT + SCROLLBAR_DIMENSION,
  SELECT_WIDTH: 48,
  MENU_WIDTH: 32 + (20 - SCROLLBAR_DIMENSION),

  // TODO: this is componentSize, but that one is a string and contains 'px'
  ROW_HEIGHT: 56,
}
