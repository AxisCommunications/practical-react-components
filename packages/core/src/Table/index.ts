export { Table } from './Table'
export { TableRow } from './TableRow'
export { TableHeader } from './TableHeader'
export { TableHeaderText } from './TableCells'
export { TableDivider } from './TableDivider'
export { SkeletonTableRows } from './SkeletonRows'
