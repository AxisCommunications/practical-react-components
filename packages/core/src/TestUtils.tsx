import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { PracticalProvider } from './Practical'

/*
  Only used for Jest Snapshot Testing
*/

export const TP = ({
  children,
}: {
  readonly children: React.ReactElement
}): JSX.Element => <PracticalProvider>{children}</PracticalProvider>

/**
 * Mock reference to practical-root
 */
const createNodeMock = (element: React.ReactElement) => {
  if (element.props.id !== 'practical-root') {
    return null
  }

  return {}
}

export const TestRender = (jsx: JSX.Element) => {
  return renderer.create(<TP>{jsx}</TP>, { createNodeMock }).toJSON()
}

export const TestText = (): JSX.Element => <span>Test</span>

/* eslint-disable-next-line react/display-name */
export const TestTextWithForwardRef = React.forwardRef<HTMLElement>(
  ({ ...props }, ref) => (
    <span {...props} ref={ref}>
      Test
    </span>
  )
)
