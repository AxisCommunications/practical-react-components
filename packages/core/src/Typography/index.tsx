import styled, { css } from 'styled-components'

export type TypographyVariant =
  | 'chip-tag-text'
  | 'explanatory-text'
  | 'caption'
  | 'navigation-group'
  | 'navigation-label'
  | 'card-title'
  | 'group-title'
  | 'column-heading'
  | 'header-title'
  | 'default-text'
  | 'dialog-heading'
  | 'page-heading'
  | 'button-text'
  | 'compact-label'

const TYPOGRAPHY_VARIANTS = {
  /**
   * Paragraph
   *
   * The sort of text that appears in blocks inside
   * e.g. a modal, should be easily readable.
   */
  'default-text': {
    as: 'p',
    css: css`
      font-size: ${({ theme }) => theme.font.size.regular};
      line-height: ${({ theme }) => theme.font.lineHeight.large};
    `,
  },

  /**
   * Descriptive
   *
   * The sort of text that elaborates on another
   * piece of information or context.
   * Should be small to be able to show a lot.
   */
  caption: {
    as: 'p',
    css: css`
      font-size: ${({ theme }) => theme.font.size.small};
      line-height: ${({ theme }) => theme.font.lineHeight.small};
    `,
  },

  /**
   * Dialog heading
   */
  'dialog-heading': {
    as: 'p',
    css: css`
      font-size: ${({ theme }) => theme.font.size.larger};
      line-height: ${({ theme }) => theme.font.lineHeight.larger};
    `,
  },

  /**
   * Page heading
   */
  'page-heading': {
    as: 'p',
    css: css`
      font-size: ${({ theme }) => theme.font.size.larger};
      line-height: ${({ theme }) => theme.font.lineHeight.larger};
    `,
  },

  /**
   * Chip or tag text
   *
   * Text appearing as chip or tag.
   * Should be small text.
   */
  'chip-tag-text': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.small};
      line-height: ${({ theme }) => theme.font.lineHeight.smaller};
      white-space: nowrap;
    `,
  },

  /**
   * Explanatory text
   *
   * Text appearing as explanatory text, e.g. error message in a field.
   * Should be small text.
   */
  'explanatory-text': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.small};
      line-height: ${({ theme }) => theme.font.lineHeight.small};
      white-space: nowrap;
    `,
  },

  /**
   * Navigation group
   *
   * Text appearing as navigation group name, e.g. Group name on sidebar.
   */
  'navigation-group': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.small};
      line-height: ${({ theme }) => theme.font.lineHeight.smaller};
      white-space: nowrap;
    `,
  },

  /**
   * Navigation label
   *
   * Standard size used for text labels,
   * short descriptions connected to another element.
   * e.g. Label on sidebar
   */
  'navigation-label': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.regular};
      line-height: ${({ theme }) => theme.font.lineHeight.small};
      white-space: nowrap;
    `,
  },

  /**
   * Card title
   */
  'card-title': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.regular};
      font-weight: ${({ theme }) => theme.font.fontWeight.semibold};
      line-height: ${({ theme }) => theme.font.lineHeight.small};
      white-space: nowrap;
    `,
  },

  /**
   * Group title
   *
   * Text appearing as a title of some fields or some group in Dialog.
   */
  'group-title': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.regular};
      font-weight: ${({ theme }) => theme.font.fontWeight.semibold};
      line-height: ${({ theme }) => theme.font.lineHeight.small};
      white-space: nowrap;
    `,
  },

  /**
   * Column heading in a table
   */
  'column-heading': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.small};
      font-weight: ${({ theme }) => theme.font.fontWeight.normal};
      line-height: ${({ theme }) => theme.font.lineHeight.smaller};
      white-space: nowrap;
    `,
  },

  /**
   * Header title
   */
  'header-title': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.regular};
      font-weight: ${({ theme }) => theme.font.fontWeight.semibold};
      line-height: ${({ theme }) => theme.font.lineHeight.small};
      white-space: nowrap;
    `,
  },

  /**
   * Button
   *
   * Text appearing inside a button
   */
  'button-text': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.large};
      font-weight: ${({ theme }) => theme.font.fontWeight.semibold};
      line-height: ${({ theme }) => theme.font.lineHeight.large};
      white-space: nowrap;
    `,
  },

  /**
   * Compact label
   *
   * Label where space is limited
   */
  'compact-label': {
    as: 'span',
    css: css`
      font-size: ${({ theme }) => theme.font.size.smaller};
      line-height: ${({ theme }) => theme.font.lineHeight.smallest};
    `,
  },
}

/**
 *
 * Generic typography
 *
 */

export interface TypographyProps {
  readonly variant?: TypographyVariant
}

export const Typography = styled.p.attrs<TypographyProps>(
  ({ variant = 'default-text' }) => ({
    as: TYPOGRAPHY_VARIANTS[variant].as,
  })
)<TypographyProps>`
  font-family: ${({ theme }) => theme.font.family};
  font-size: ${({ theme }) => theme.font.size.regular};
  line-height: ${({ theme }) => theme.font.lineHeight.large};
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0;

  ${({ variant = 'default-text' }) => TYPOGRAPHY_VARIANTS[variant].css}
`
