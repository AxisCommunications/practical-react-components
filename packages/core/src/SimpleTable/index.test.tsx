import React from 'react'
import 'jest-styled-components'

import { TestRender } from '../TestUtils'
import { Typography } from '../Typography'
import { SimpleTable, SimpleTableRow } from '.'
import { TableHeaderText } from '../Table'

const clickHandler = () => console.log('Row click')

const TABLE_HEADER_DATA = [
  { id: 'serial-number', title: 'Serial number' },
  { id: 'name', title: 'Name' },
  { id: 'ip-address', title: 'IP address' },
]

const DEVICE_LIST = [
  { serialNumber: 'FACEFACE0001', name: 'a', ip: '192.168.0.1' },
  { serialNumber: 'FACEFACE0002', name: 'b', ip: '192.168.0.2' },
  { serialNumber: 'FACEFACE0003', name: 'c', ip: '192.168.0.3' },
  { serialNumber: 'FACEFACE0004', name: 'd', ip: '192.168.0.4' },
  { serialNumber: 'FACEFACE0005', name: 'e', ip: '192.168.0.5' },
  { serialNumber: 'FACEFACE0006', name: 'f', ip: '192.168.0.6' },
  { serialNumber: 'FACEFACE0007', name: 'g', ip: '192.168.0.7' },
  { serialNumber: 'FACEFACE0008', name: 'h', ip: '192.168.0.8' },
  { serialNumber: 'FACEFACE0009', name: 'i', ip: '192.168.0.9' },
  { serialNumber: 'FACEFACE0010', name: 'j', ip: '192.168.0.10' },
  { serialNumber: 'FACEFACE0011', name: 'k', ip: '192.168.0.11' },
  { serialNumber: 'FACEFACE0012', name: 'l', ip: '192.168.0.12' },
]

describe('SimpleTable', () => {
  test('SimpleTable', () => {
    const tree = TestRender(
      <SimpleTable
        maxHeight={3}
        header={TABLE_HEADER_DATA.map(({ title }, id) => (
          <TableHeaderText key={id}>{title}</TableHeaderText>
        ))}
        className="my-simple-table"
      >
        {DEVICE_LIST.map((device, index) => (
          <SimpleTableRow
            key={device.serialNumber}
            disabled={index === 2}
            className="my-simple-table-row"
            onClick={clickHandler}
          >
            <Typography>{device.serialNumber}</Typography>
            <Typography>{device.name}</Typography>
            <Typography>{device.ip}</Typography>
          </SimpleTableRow>
        ))}
      </SimpleTable>
    )
    expect(tree).toMatchSnapshot()
  })
})
