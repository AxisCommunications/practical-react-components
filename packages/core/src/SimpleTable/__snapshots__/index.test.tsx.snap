// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleTable SimpleTable 1`] = `
.c6 {
  font-family: Open Sans,Helvetica,arial,sans-serif;
  font-size: 13px;
  line-height: 22px;
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0;
  font-size: 12px;
  font-weight: 400;
  line-height: 17px;
  white-space: nowrap;
}

.c11 {
  font-family: Open Sans,Helvetica,arial,sans-serif;
  font-size: 13px;
  line-height: 22px;
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0;
  font-size: 13px;
  line-height: 22px;
}

.c14 {
  position: fixed;
  top: 0;
  width: 360px;
  border-radius: 4px;
  margin: 8px 16px;
  right: 0;
}

.c0 {
  color: rgb(41,41,41);
}

.c1 {
  position: absolute;
  z-index: 0;
}

.c13 {
  position: absolute;
  z-index: 1;
}

.c4 {
  height: 48px;
  text-align: left;
  display: table;
  table-layout: fixed;
  width: 100%;
}

.c5 {
  color: rgb(61,61,61);
  background: rgb(255,255,255);
  padding-right: 16px;
  text-align: left;
  font-weight: 400;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c5:first-child {
  padding-left: 24px;
}

.c9 {
  display: table;
  table-layout: fixed;
  width: 100%;
  height: 56px;
  text-align: left;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  cursor: pointer;
}

.c9:nth-child(odd) {
  background-color: rgb(250,250,250);
}

.c9:nth-child(even) {
  background-color: rgb(255,255,255);
}

.c9:hover {
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  z-index: 1;
}

.c12 {
  display: table;
  table-layout: fixed;
  width: 100%;
  height: 56px;
  text-align: left;
  -webkit-transition: box-shadow 200ms;
  transition: box-shadow 200ms;
  opacity: 0.48;
  pointer-events: none;
  cursor: unset;
}

.c12:nth-child(odd) {
  background-color: rgb(250,250,250);
}

.c12:nth-child(even) {
  background-color: rgb(255,255,255);
}

.c12:hover {
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  z-index: 1;
}

.c10 {
  padding-right: 16px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c10:first-child {
  padding-left: 24px;
}

.c2 {
  border-collapse: collapse;
  width: 100%;
  height: 100%;
  overflow-y: auto;
}

.c3 {
  display: table;
  table-layout: fixed;
  width: 100%;
  border-bottom: 1px solid rgb(230,230,230);
}

.c8 {
  display: block;
  width: 100%;
  max-height: 168px;
  overflow: auto;
  padding-bottom: 8px;
}

.c7 {
  text-transform: none;
}

<div
  className="c0"
  id="practical-root"
>
  <div>
    <div
      className="c1"
      id="layer-1"
    >
      <table
        className="c2 my-simple-table"
      >
        <thead
          className="c3"
        >
          <tr
            className="c4"
          >
            <th
              className="c5"
            >
              <p
                className="c6 c7"
              >
                Serial number
              </p>
            </th>
            <th
              className="c5"
            >
              <p
                className="c6 c7"
              >
                Name
              </p>
            </th>
            <th
              className="c5"
            >
              <p
                className="c6 c7"
              >
                IP address
              </p>
            </th>
          </tr>
        </thead>
        <tbody
          className="c8"
        >
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0001
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                a
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.1
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0002
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                b
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.2
              </p>
            </td>
          </tr>
          <tr
            className="c12 my-simple-table-row"
            disabled={true}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0003
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                c
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.3
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0004
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                d
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.4
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0005
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                e
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.5
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0006
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                f
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.6
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0007
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                g
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.7
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0008
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                h
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.8
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0009
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                i
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.9
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0010
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                j
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.10
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0011
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                k
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.11
              </p>
            </td>
          </tr>
          <tr
            className="c9 my-simple-table-row"
            disabled={false}
            onClick={[Function]}
          >
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                FACEFACE0012
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                l
              </p>
            </td>
            <td
              className="c10"
            >
              <p
                className="c11"
              >
                192.168.0.12
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div>
    <div
      className="c13"
      id="layer-2"
    >
      <div
        className="c14"
      />
    </div>
  </div>
</div>
`;
